From a32fc3bc4b6bf2e9a564b409dfbb605ee742c52a Mon Sep 17 00:00:00 2001
From: Tejas Patel <tejas.patel@xilinx.com>
Date: Wed, 17 Jun 2020 05:18:05 -0700
Subject: [PATCH 1383/1851] xilinx: firmware: Add sysfs to get last reset
 reason

commit 2fc4036ac1b2ba41583e18ad740db5bbeba75af4 from
https://github.com/Xilinx/linux-xlnx.git

Add sysfs to get last reset reason.

Signed-off-by: Tejas Patel <tejas.patel@xilinx.com>
Reviewed-by: Jolly Shah <jolly.shah@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 .../ABI/stable/sysfs-firmware-zynqmp          | 13 +++++++
 drivers/firmware/xilinx/zynqmp.c              | 36 +++++++++++++++++++
 include/linux/firmware/xlnx-zynqmp.h          | 11 ++++++
 3 files changed, 60 insertions(+)

diff --git a/Documentation/ABI/stable/sysfs-firmware-zynqmp b/Documentation/ABI/stable/sysfs-firmware-zynqmp
index eeae291a048c..c43479fccb1d 100644
--- a/Documentation/ABI/stable/sysfs-firmware-zynqmp
+++ b/Documentation/ABI/stable/sysfs-firmware-zynqmp
@@ -101,3 +101,16 @@ Description:
 		# echo 0 > /sys/firmware/zynqmp/health_status
 
 Users:		Xilinx
+
+What:		/sys/firmware/zynqmp/last_reset_reason
+Date:		June 2020
+KernelVersion:	5.4.0
+Contact:	"Tejas Patel" <tejasp@xilinx.com>
+Description:
+		This sysfs interface allows to get last reset reason.
+
+		Usage:
+		Get last reset reason
+		# cat /sys/firmware/zynqmp/last_reset_reason
+
+Users:		Xilinx
diff --git a/drivers/firmware/xilinx/zynqmp.c b/drivers/firmware/xilinx/zynqmp.c
index 95a94aaa8cb1..1883e57d2e7e 100644
--- a/drivers/firmware/xilinx/zynqmp.c
+++ b/drivers/firmware/xilinx/zynqmp.c
@@ -1607,10 +1607,46 @@ static ssize_t config_reg_show(struct kobject *kobj,
 static struct kobj_attribute zynqmp_attr_config_reg =
 					__ATTR_RW(config_reg);
 
+static ssize_t last_reset_reason_show(struct kobject *kobj,
+				      struct kobj_attribute *attr,
+				      char *buf)
+{
+	int ret;
+	u32 ret_payload[PAYLOAD_ARG_CNT];
+
+	ret = zynqmp_pm_get_last_reset_reason(ret_payload);
+	if (ret)
+		return ret;
+	switch (ret_payload[1]) {
+	case PM_RESET_REASON_EXT_POR:
+		return sprintf(buf, "ext_por\n");
+	case PM_RESET_REASON_SW_POR:
+		return sprintf(buf, "sw_por\n");
+	case PM_RESET_REASON_SLR_POR:
+		return sprintf(buf, "sl_por\n");
+	case PM_RESET_REASON_ERR_POR:
+		return sprintf(buf, "err_por\n");
+	case PM_RESET_REASON_DAP_SRST:
+		return sprintf(buf, "dap_srst\n");
+	case PM_RESET_REASON_ERR_SRST:
+		return sprintf(buf, "err_srst\n");
+	case PM_RESET_REASON_SW_SRST:
+		return sprintf(buf, "sw_srst\n");
+	case PM_RESET_REASON_SLR_SRST:
+		return sprintf(buf, "slr_srst\n");
+	default:
+		return sprintf(buf, "unknown reset\n");
+	}
+}
+
+static struct kobj_attribute zynqmp_attr_last_reset_reason =
+					__ATTR_RO(last_reset_reason);
+
 static struct attribute *attrs[] = {
 	&zynqmp_attr_shutdown_scope.attr,
 	&zynqmp_attr_health_status.attr,
 	&zynqmp_attr_config_reg.attr,
+	&zynqmp_attr_last_reset_reason.attr,
 	NULL,
 };
 
diff --git a/include/linux/firmware/xlnx-zynqmp.h b/include/linux/firmware/xlnx-zynqmp.h
index 9b8ba835cf13..3b0ec22aaa3e 100644
--- a/include/linux/firmware/xlnx-zynqmp.h
+++ b/include/linux/firmware/xlnx-zynqmp.h
@@ -548,6 +548,17 @@ enum pm_node_id {
 	NODE_MAX,
 };
 
+enum pm_reset_reason {
+	PM_RESET_REASON_EXT_POR = 0,
+	PM_RESET_REASON_SW_POR = 1,
+	PM_RESET_REASON_SLR_POR = 2,
+	PM_RESET_REASON_ERR_POR = 3,
+	PM_RESET_REASON_DAP_SRST = 7,
+	PM_RESET_REASON_ERR_SRST = 8,
+	PM_RESET_REASON_SW_SRST = 9,
+	PM_RESET_REASON_SLR_SRST = 10,
+};
+
 /**
  * struct zynqmp_pm_query_data - PM query data
  * @qid:	query ID
-- 
2.31.1

